**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural** 

***
Datos del alumno (Nombre y Apellidos):

Fecha:
***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Named-Entity Recognition</span>

**Objetivos** 

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

Observa las diferentes librerías que se están importando.

In [1]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
import es_core_news_md

El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>: 

https://spacy.io/models/es#es_core_news_md


In [2]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [3]:
filename = "./02Dataset_anonimizado.csv"
lines_number = 10000 # limite de carga para analisis
#datdata = pd.read_csv(filename, delimiter=';',nrows=lines_number)  a = pd.read_csv(filename, delimiter=';')  
encoding='latin-1'
data = pd.read_csv(filename, delimiter=';', nrows=lines_number, encoding=encoding)

El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas: 


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos 

In [4]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
#print(data["CONTENIDO A ANALIZAR"][1])
#print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, lines_number):
    
    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)
    
    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


#ejemplo de cómo recorrer un comentario palabra por palabra    
""" for token in doc[1]:
    print(token) """
#Documentacion de los atributtos que ofrece la libreria spacy con el modelo para obtener información, de la cadena entrante
    
""" for token in doc[1]:
    print(token.text)        # texto original:      "producto"
    print(token.lemma_)      # forma base/lema:     "producto"
    print(token.pos_)        # categoría gramatical: "NOUN"
    print(token.tag_)        # etiqueta específica detallada: "VMP00SM"
    print(token.dep_)        # dependencia sintáctica: "nsubj"
    print(token.is_stop)     # ¿es palabra vacía?:  False
    print(token.is_punct)    # ¿es puntuación?:     False    
    print(token.lower_)      # texto en minúsculas:           "corriendo"
    print(token.norm_)       # forma normalizada:             "corriendo"
    print(token.shape_)      # patrón de forma:               "Xxxxx"
    print(token.pos_)        # categoría universal (UPOS):    "VERB"
    print(token.tag_)        # etiqueta específica detallada: "VMP00SM"
    print(token.explain(token.tag_))  # explicación del tag   "verb participle"
    print(token.head.text)   # token del que depende:         "corre"
    print(token.is_punct)    # ¿es puntuación?:   True/False
    print(token.is_space)    # ¿es espacio?:       True/False
    print(token.is_alpha)    # ¿solo letras?:      True/False
    print(token.is_digit)    # ¿es número?:        True/False
    print(token.is_upper)    # ¿está en mayúscula?: True/False
    print(token.is_lower)    # ¿está en minúscula?: True/False
    print(token.like_email)  # ¿parece email?:     True/False
    print(token.like_url)    # ¿parece URL?:       True/False
    print(token.like_num)    # ¿parece número?:    True/False  ("veinte", "20")
    print(token.ent_type_)   # tipo de entidad:   "PER", "LOC", "ORG"
    print(token.ent_iob_)    # posición en entidad: "B"=inicio, "I"=dentro, "O"=fuera """
    
filas = []
for token in doc[1]:
    filas.append({
        "texto":    token.text,
        "lower":    token.lower_,
        "lemma":    token.lemma_,
        "pos":      token.pos_,
        "tag":      token.tag_,
        "dep":      token.dep_,
        "head":     token.head.text,
        "is_stop":  token.is_stop,
        "is_punct": token.is_punct,
        "is_alpha": token.is_alpha,
        "entidad":  token.ent_type_
    })

df_tokens = pd.DataFrame(filas)
print(df_tokens)   

       texto    lower    lemma    pos    tag     dep     head  is_stop  \
0         el       el       el    DET    DET     det     real     True   
1       real     real     real  PROPN  PROPN   nsubj   puesto    False   
2     madrid   madrid   madrid  PROPN  PROPN    flat     real    False   
3         ha       ha    haber    AUX    AUX     aux   puesto     True   
4     puesto   puesto    poner   VERB   VERB    ROOT   puesto    False   
..       ...      ...      ...    ...    ...     ...      ...      ...   
208      del      del      del    ADP    ADP    case  partido     True   
209  partido  partido  partido   NOUN   NOUN    nmod   minuto    False   
210       en       en       en    ADP    ADP  advmod  partido     True   
211  directo  directo  directo   NOUN   NOUN   fixed       en    False   
212        :        :        :  PUNCT  PUNCT   punct  contado    False   

     is_punct  is_alpha entidad  
0       False      True          
1       False      True     ORG  
2       F

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

In [5]:
# Incluye aquí el código generado para poder responder a tu pregunta
len_corpus = len(data)
lenfilas_corpus = data.shape[0] 
len_columnas_corpus = data.columns
len_filas_columnas_corpus = data.shape
print(f"Registros cargados: {len_corpus}")
print(f"Columnas: {list(len_columnas_corpus)}")
print(f" filas y columnas: {len_filas_columnas_corpus}")



Registros cargados: 10000
Columnas: ['MEDIO', 'SOPORTE', 'URL', 'TIPO DE MENSAJE', 'CONTENIDO A ANALIZAR', 'INTENSIDAD', 'TIPO DE ODIO', 'TONO HUMORISTICO', 'MODIFICADOR', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15']
 filas y columnas: (10000, 16)


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<p>
Podemos usar funciones de conteo de arrays de datos en python como len, o asi mismo explorando la documentación de la libreria padas cada dataframe nos indica el tamaño de este a nivel de filas
</p>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [6]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Una vez que ya tienes la lista doc procesada
total_tokens = 0
total_tokens_sin_stopwords = 0
total_tokens_sin_puntuacion = 0
#documento es el listado de palabras separas en tokens
for documento in doc:
    total_tokens += len(documento) # suma cada palabra 
    total_tokens_sin_stopwords += len([t for t in documento if not t.is_stop]) # contabiliza si la palabra NO es vacia 
    total_tokens_sin_puntuacion += len([t for t in documento if not t.is_punct]) # contabiliza si la palabra NO tiene puntuacion

print(f"Total tokens (con todo):              {total_tokens}")
print(f"Total tokens (sin stopwords):         {total_tokens_sin_stopwords}")
print(f"Total tokens (sin puntuación):        {total_tokens_sin_puntuacion}")

Total tokens (con todo):              252849
Total tokens (sin stopwords):         134923
Total tokens (sin puntuación):        219179


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<p> Tengamos en cuenta que la libreria de spacy ya cuenta con un modelo que le permite generar una normalización de los datos preparando el flujo de entrada antes de un analisis, en donde primero ejecutamos la fase de tokenizacion sobre la columna CONTENIDO A ANALIZAR de nuestro dataframe, alli ya tenemos la frase separada es decir las palabras simplemente debemos interar en dicho array con los n tokens y contar, adicional se usaron dos funciones adicionales que a mi manera sirven para generar unas estadisticas sobre la informacion a ser procesada.

</p>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [7]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Calcular número de palabras por cada comentario
data["num_palabras"] = data["CONTENIDO A ANALIZAR"].str.split().str.len()
print(data["num_palabras"])
# Promedio
promedio = data["num_palabras"].mean()
print(f"Promedio de palabras por comentario: {promedio:.2f}")

0        25
1       195
2       127
3        19
4       147
       ... 
9995      2
9996     15
9997      4
9998      3
9999     10
Name: num_palabras, Length: 10000, dtype: int64
Promedio de palabras por comentario: 21.68


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<p>Aqui preferí usar pandas, ejecutando los siguientes pasos: a.para leer y contar las palabras que tenia cada fila del comnetario. b.despues de tener el numero de palabras, pude calcular el promedio de estas usando la función mean() de pandas, tambien podia hacerlo con la libreria spacy pero vi menos limpio el codigo. </p>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [8]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Clasificar comentarios según la regla de intensidad

data["num_palabras"] = data["CONTENIDO A ANALIZAR"].str.split().str.len()
data["contiene_odio"] = data["INTENSIDAD"].apply(lambda x: "Odio" if x > 0 else "No odio")

# Promedio de palabras por categoría
promedio_por_categoria = data.groupby("contiene_odio")["num_palabras"].agg(
    promedio="mean",
    total_comentarios="count",
    total_palabras="sum",
    maximo="max",
    minimo="min"
).round(2)

print(promedio_por_categoria)

               promedio  total_comentarios  total_palabras  maximo  minimo
contiene_odio                                                             
No odio           21.58               9252          199621    2809       1
Odio              22.91                748           17140    1990       1


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<p>Reutilice la función de conteo de palabras para generar un total, por fila como primer punto, ahora segmente la regla de que comentarios tienen odio de acuerdo a lo descripto en la actividad en donde si es > 0 existe odio, basandose en estos dos grupos de datos, use el principio sql groupby para agrupar, la totalización en 2 grupos con odio y sin odio.</p> 
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [9]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Contar oraciones por cada comentario
num_oraciones  = []
oraciones_list = []

for documento in doc:
    oraciones = list(documento.sents) # segmentacion de oraciones
    num_oraciones.append(len(oraciones))
    oraciones_list.append([s.text for s in oraciones])

# Agregar al dataframe
data["num_oraciones"] = num_oraciones
data["oraciones"]     = oraciones_list

#print(data[["CONTENIDO A ANALIZAR", "num_oraciones", "INTENSIDAD"]].head(10))

#print(data["num_oraciones"].describe().round(2))

# Por intensidad
print(data.groupby("contiene_odio")["num_oraciones"].agg(
    promedio="mean",
    maximo="max",
    minimo="min",
    total_comentarios="count"
).round(2))

               promedio  maximo  minimo  total_comentarios
contiene_odio                                             
No odio            1.78      79       1               9252
Odio               1.77      42       1                748


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<p>Aqui explore la funcion de la libreria spacy de segmentacion de oraciones en donde a la columna de comentario le obtuvimos el numero de oraciones y los segmentamos para que en una misma fila podemos tener n oraciones, de acuerdo a la naturaleza del dato, y nuevamente realizamos un group by en donde las estamos agrupando por la condición del dataset por odio y no odio </p>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

In [10]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Extraer entidades de cada comentario
from collections import Counter

entidades_odio    = []
entidades_no_odio = []
for i, documento in enumerate(doc):
    
    if documento.ents:  # si encontró entidades
        for entidad in documento.ents:
            if data['INTENSIDAD'][i] > 0:
                entidades_odio.append({"texto":  entidad.text.lower(),
                "tipo":   entidad.label_})
            else:
                entidades_no_odio.append({"texto":  entidad.text.lower(),
                "tipo":   entidad.label_})              

# Convertir a dataframe para analizar
df_odio    = pd.DataFrame(entidades_odio)
df_no_odio = pd.DataFrame(entidades_no_odio)
#Estas metricas permiten establcer la cantidad de entidades que hay en cada comentario con odio y sin odio y saber cuales se repiten encontre mucho la palabra de groseria y trump
#print("=== Top entidades CON odio ===", len(df_odio))
#print(df_odio["texto"].value_counts().head(10))

#print("\n=== Top entidades SIN odio ===", len(df_no_odio))
#print(df_no_odio["texto"].value_counts().head(10))

# También por tipo
#print("\n=== Tipos de entidad en comentarios CON odio ===")
#print(df_odio["tipo"].value_counts())
#Esta es una opcion para sacar el porcentaje usando funciones manuales, pero Pandas al convertir esto en un dataframe, nos da la opcion de cruzar datos

data["tiene_ner"] = [len(documento.ents) > 0 for documento in doc]
data["grupo"]     = data["INTENSIDAD"].apply(lambda x: "Odio" if x > 0 else "No odio")

# Tabla cruzada con porcentajes
tabla = pd.crosstab(
    data["grupo"],
    data["tiene_ner"],
    colnames=["Tiene NER"],
    margins=True,
    normalize="index"  # porcentaje por fila
) * 100
# Renombrar columnas para que sea legible
tabla.columns     = ["Sin entidades NER (%)", "Con entidades NER (%)"]
tabla.index.name  = "Grupo"

# Renombrar fila All
tabla = tabla.rename(index={True: "Total corpus"})
print(tabla.round(2))

         Sin entidades NER (%)  Con entidades NER (%)
Grupo                                                
No odio                  66.09                  33.91
Odio                     62.43                  37.57
All                      65.82                  34.18


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b> <p> Aqui volvi a usar pandas primero quise explorar que es la funcion de document.ents para asi entender cual es su retorno, en donde es muy interesante ver que podemos obtener de la entidad vs el mundo real, si se refiere a una persona, pais institucion o otro tipo de entidad y una descripcion lo cual me parece muy util para la clasificacion y analisis de los comentarios, seguido a esto aplique las reglas en donde se cuenta el valor si es una entidad NER y clasifico si estan dentro del grupo con odio o no, seguido a esto genere una tabla de frecuencia de pandas que me permite hacer cruces con variables de categoria, para poder calcular el porcentaje se multiplico * 100 y seguido a esto obtenemos el porcentaje de comentarios que cumplen con las condiciones del enunciado, en donde podemos inferir que de la categorizacion el 33.91 %  contiene una entidad NER lo cual indica que el 66.09 % hace referencia a un concepto abstracto, en donde no hay una forma en la oración que se adapte a la forma. </p>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 7.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER de tipo PERSON en cada grupo?</span>

In [11]:
entidades_odio    = []
entidades_no_odio = []
entidad_person = 'PER'
count_odio_con_per    = 0
count_odio_sin_per    = 0
count_no_odio_con_per = 0
count_no_odio_sin_per = 0
count_total_odio = 0
count_total_no_odio = 0
for i, documento in enumerate(doc):
    
    if documento.ents:  # si encontró entidades
        for entidad in documento.ents:
            
                if data['INTENSIDAD'][i] > 0:
                    entidades_odio.append({"texto":  entidad.text.lower(),
                    "tipo":   entidad.label_})
                    count_odio_con_per += 1 if entidad.label_ == entidad_person else 0
                    count_odio_sin_per += 1 if entidad.label_ != entidad_person else 0
                    
                else:
                    entidades_no_odio.append({"texto":  entidad.text.lower(),
                    "tipo":   entidad.label_})
                    count_no_odio_con_per += 1 if entidad.label_ == entidad_person else 0
                    count_no_odio_sin_per += 1 if entidad.label_ != entidad_person else 0

                    

# Convertir a dataframe para analizar

df_odio    = pd.DataFrame(entidades_odio)
df_no_odio = pd.DataFrame(entidades_no_odio)
count_total_odio = len(df_odio)
count_total_no_odio = len(df_no_odio)

print(f"Personas en comentarios CON odio:    {count_total_odio}")
print(f"Personas en comentarios SIN odio:    {count_total_no_odio}")

# Top personas más mencionadas con porcentaje
print("\n=== Personas más mencionadas en comentarios CON odio ===")

print(df_odio[df_odio["tipo"] == entidad_person]["texto"].value_counts().head(10))

print("\n=== Personas más mencionadas en comentarios SIN odio ===")
print(df_no_odio[df_no_odio["tipo"] == entidad_person]["texto"].value_counts().head(10))

resumen = pd.DataFrame({
    "Grupo":          ["CON odio",     "SIN odio"],
    "Total":          [count_total_odio,      count_total_no_odio],
    "Con PER (n)":    [count_odio_con_per,    count_no_odio_con_per],
    "Sin PER (n)":    [count_odio_sin_per,    count_no_odio_sin_per],
    "Con PER (%)":    [round(count_odio_con_per    / count_total_odio    * 100, 2),
                       round(count_no_odio_con_per / count_total_no_odio * 100, 2)],
    "Sin PER (%)":    [round(count_odio_sin_per    / count_total_odio    * 100, 2),
                       round(count_no_odio_sin_per / count_total_no_odio * 100, 2)]
}).set_index("Grupo")

print(resumen)




Personas en comentarios CON odio:    654
Personas en comentarios SIN odio:    7080

=== Personas más mencionadas en comentarios CON odio ===
texto
puta                           12
trump                           7
més                             3
ã‚â                             3
gentuza                         3
cela                            2
aquest suplement                2
dã¢â€â™histí²ria contemporí     2
jordi pujol i                   2
doncs                           2
Name: count, dtype: int64

=== Personas más mencionadas en comentarios SIN odio ===
texto
trump           180
biden            51
messi            49
jajajaja         24
donald trump     20
hi               19
covid19          16
yoko ono         16
bueno            15
més              14
Name: count, dtype: int64
          Total  Con PER (n)  Sin PER (n)  Con PER (%)  Sin PER (%)
Grupo                                                              
CON odio    654          290          364        44.34       

<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
<p>Que realice aqui, me base en el ciclo for de todos los tokens y entidades de mi dataset, en donde les coloque contadores conforme, a la necesidad totales y el filtro, que indica cuales son entidades igual a PER y cuales NO segmentado con sin son del grupo ODIO o no ODIO es una función sencilla que usa contadores y luego con ellos calcula los porcentajes, de este podemos analizar a nivel de porcentajes que las palabras con ODIO en su mayoria no se refieren a una entidad de tipo persona, y otro print que realice hace referencia al nombre trump, en muchas coincidencias  </p>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 8.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de palabras en cada combinación posible de género y número (p.ej. masculino singular) en cada grupo?</span>

In [12]:
# Incluye aquí el código generado para poder responder a tu pregunta
grupo_odio_entidad    = []
grupo_no_odio_entidad = []
total_tokens=0
for i, documento in enumerate(doc):
    #print(f"Comentario {i+1}: {data['CONTENIDO A ANALIZAR'][i]}")
    if data['INTENSIDAD'][i] > 0:
        #Interactuar con cada token del comentario donde el valor de intensidad es mayor a 0 es decir con ODIO
        for token in documento:
            genero = token.morph.get("Gender")  # devuelve lista ['Masc'] o []
            numero = token.morph.get("Number")  # devuelve lista ['Plur'] o []
            #Combinaciones posibles de genero y numero: Masculino singular, Masculino plural, Femenino singular, Femenino plural, Neutro singular, Neutro plural
            #print(f"Token: {token.text}, Género: {genero}, Número: {numero}")
            grupo_odio_entidad.append((token.text, genero, numero))
            total_tokens += 1
            
    else:
        #Interactuar con cada token del comentario donde el valor de intensidad es igual a 0 es decir sin ODIO
        for token in documento:
            genero = token.morph.get("Gender")  # devuelve lista ['Masc'] o []
            numero = token.morph.get("Number")  # devuelve lista ['Plur'] o []
           # print(f"Token: {token.text}, Género: {genero}, Número: {numero}")
            grupo_no_odio_entidad.append((token.text, genero, numero))  
            total_tokens += 1
# Convertir listas a dataframe
df_odio = pd.DataFrame(grupo_odio_entidad, columns=["token", "genero", "numero"])
df_no_odio = pd.DataFrame(grupo_no_odio_entidad, columns=["token", "genero", "numero"])

# Limpiar listas a strings simples (vienen como ['Masc'] o [])
df_odio["genero"]    = df_odio["genero"].apply(lambda x: x[0] if x else "N/A")
df_odio["numero"]    = df_odio["numero"].apply(lambda x: x[0] if x else "N/A")
df_no_odio["genero"] = df_no_odio["genero"].apply(lambda x: x[0] if x else "N/A")
df_no_odio["numero"] = df_no_odio["numero"].apply(lambda x: x[0] if x else "N/A")

# Combinar género y número en una sola columna
df_odio["combinacion"]    = df_odio["genero"]    + " + " + df_odio["numero"]
df_no_odio["combinacion"] = df_no_odio["genero"] + " + " + df_no_odio["numero"]

resumen_completo = pd.DataFrame({
    "CON odio (n)":  df_odio["combinacion"].value_counts(),
    "CON odio (%)":  df_odio["combinacion"].value_counts(normalize=True).mul(100).round(2),
    "SIN odio (n)":  df_no_odio["combinacion"].value_counts(),
    "SIN odio (%)":  df_no_odio["combinacion"].value_counts(normalize=True).mul(100).round(2),
}).fillna(0)

print(f"\nTotal tokens CON odio:    {len(df_odio)}")
print(f"Total tokens SIN odio:    {len(df_no_odio)}")
print(f"Total tokens corpus:      {total_tokens}")
print()
print(resumen_completo)        



Total tokens CON odio:    20080
Total tokens SIN odio:    232769
Total tokens corpus:      252849

             CON odio (n)  CON odio (%)  SIN odio (n)  SIN odio (%)
combinacion                                                        
Fem + N/A              11          0.05           128          0.05
Fem + Plur            561          2.79          6755          2.90
Fem + Sing           2123         10.57         24109         10.36
Masc + N/A             10          0.05           227          0.10
Masc + Plur          1185          5.90         12138          5.21
Masc + Sing          2736         13.63         31293         13.44
N/A + N/A           10140         50.50        118545         50.93
N/A + Plur           1108          5.52         12205          5.24
N/A + Sing           2206         10.99         27369         11.76


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
<p>
Aqui en esta solución lo que hicimos fue, obtenemos las frases, iteramos en los tokens y armamos los datasets de acuerdo a los grupos oraciones con odio y sin odio, y que hicimos generar combinaciones si es femenino el token (palabra) y si esta en plurar o no, se hizo usando la funcion token.morph.get, para ello, y en analisis del corpus vemos que existe en las frases de odio mayor referencia a un hombre que hacia una mujer.

 </p>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 9.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio), indica cuántas entidades de cada tipo posible se reconocen en cada uno de los grupos</span>

In [13]:
# Incluye aquí el código generado para poder responder a tu pregunta
grupo_odio_entidad    = []
grupo_no_odio_entidad = []
total_entidades=0
for i, documento in enumerate(doc):
    
       # print(f"\nComentario {i} (Intensidad: {data['INTENSIDAD'][i]}): ({data['CONTENIDO A ANALIZAR'][i]})")
    if data['INTENSIDAD'][i] > 0:
        if documento.ents:  # si encontró entidades
            for entidad in documento.ents:
                total_entidades += 1
                #print(f" - Entidad: '{entidad.text}' | Tipo: {entidad.label_}")
                grupo_odio_entidad.append((entidad.text, entidad.label_))
        else:
            grupo_odio_entidad.append(("sin entidad", "N/A"))     
    else:
        if documento.ents:  # si encontró entidades
            
            for entidad in documento.ents:
                #print(f" - Entidad: '{entidad.text}' | Tipo: {entidad.label_}")
                total_entidades += 1
                grupo_no_odio_entidad.append((entidad.text, entidad.label_))
        else:
            grupo_no_odio_entidad.append(("sin entidad", "N/A"))
# Convertir listas a dataframe
df_odio = pd.DataFrame(grupo_odio_entidad, columns=["entidad", "tipo"])
df_no_odio = pd.DataFrame(grupo_no_odio_entidad, columns=["entidad", "tipo"])
print(f"Total entidades CON odio:    {len(df_odio[df_odio['entidad'] != 'sin entidad'])}")
print(f"Total entidades SIN odio:    {len(df_no_odio[df_no_odio['entidad'] != 'sin entidad'])}")
print(f"Total entidades validas en corpus:   {total_entidades}")
resumen_entidades = pd.DataFrame({
    "CON odio (n)":  df_odio["tipo"].value_counts(),
    "CON odio (%)":  df_odio["tipo"].value_counts(normalize=True).mul(100).round(2),
    "SIN odio (n)":  df_no_odio["tipo"].value_counts(),
    "SIN odio (%)":  df_no_odio["tipo"].value_counts(normalize=True).mul(100).round(2),
}).fillna(0)
print(resumen_entidades)

Total entidades CON odio:    654
Total entidades SIN odio:    7080
Total entidades validas en corpus:   7734
      CON odio (n)  CON odio (%)  SIN odio (n)  SIN odio (%)
tipo                                                        
N/A            467         41.66          6115         46.34
PER            290         25.87          2946         22.33
LOC            176         15.70          1890         14.32
MISC            97          8.65          1251          9.48
ORG             91          8.12           993          7.53


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
<p> 
Aqui que hicimos, se intero sobre el dataset y se tomo las entidades, sumando las que si estan en un grupo y las que no aplican para obtener el 100 %, es una iteraccion sencilla y basada en lo que hemos propuesto antes set de datos x grupo, ahora que analizo vs mi carpus en las oraciones con odio, se refieren mayormente a una persona o una locacion.
</p>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 10.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio), extrae y muestra los 100 lemas más repetidos en los comentarios de cada grupo</span>

In [14]:
# Incluye aquí el código generado para poder responder a tu pregunta
grupo_odio_lema    = []
grupo_no_odio_lema = []
total_tokens=0
for i, documento in enumerate(doc):
    #print(f"Comentario {i+1}: {data['CONTENIDO A ANALIZAR'][i]}")
    comentario = data['CONTENIDO A ANALIZAR'][i]
    if data['INTENSIDAD'][i] > 0:
        #Interactuar con cada token del comentario donde el valor de intensidad es mayor a 0 es decir con ODIO
        for token in documento:
            total_tokens += 1
            if token.is_stop or token.is_punct or token.is_space:
                continue  # saltar stopwords y puntuación
            lema = token.lemma_
            grupo_odio_lema.append((comentario,lema))
            
            
    else:
        #Interactuar con cada token del comentario donde el valor de intensidad es igual a 0 es decir sin ODIO
        for token in documento:
            total_tokens += 1
            if token.is_stop or token.is_punct or token.is_space:
                continue  # saltar stopwords y puntuación
            lema = token.lemma_
            grupo_no_odio_lema.append((comentario,lema))
# Convertir listas a dataframe
df_odio = pd.DataFrame(grupo_odio_lema, columns=["comentario", "lema"])
df_no_odio = pd.DataFrame(grupo_no_odio_lema, columns=["comentario", "lema"])
# Top 100 lemas más repetidos por grupo
top100_odio    = df_odio["lema"].value_counts().head(100)
top100_no_odio = df_no_odio["lema"].value_counts().head(100)

print("=== Top 100 lemas CON odio ===")
print(top100_odio)

print("\n=== Top 100 lemas SIN odio ===")
print(top100_no_odio)

    


=== Top 100 lemas CON odio ===
lema
USUARIO_OFUSCADO    93
mierda              58
puta                52
hijo                47
i                   40
                    ..
llamar               9
tomar                9
permitir             9
resto                9
social               9
Name: count, Length: 100, dtype: int64

=== Top 100 lemas SIN odio ===
lema
USUARIO_OFUSCADO    1765
i                    707
gobierno             338
pasar                327
gente                292
                    ... 
ciudadano             87
único                 86
elección              86
impuesto              86
derecha               85
Name: count, Length: 100, dtype: int64


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
<p>
En el orden del algoritmo se itero el dataset, luego los tokens y de esto por cada palabra se extrajo el lema, y se añadio a las listas filtradas con porcentaje de odio y no odio, seguido a esto se realizo con pandas cada data set, ya con las caracteristicas previas y se determino, cual era el top, en donde los lemas mas usados en oraciones con odio son groserias.
 </p>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 11.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

In [35]:
def definir_diccionario_lemas_odio(df_odio, df_no_odio):
    # Aquí puedes definir tu diccionario de lemas relacionados con el odio
    freq_odio    = df_odio["lema"].value_counts(normalize=True).mul(100).round(4)
    freq_no_odio = df_no_odio["lema"].value_counts(normalize=True).mul(100).round(4)

    diccionario_odio = pd.DataFrame({
        "frec_odio (%)":    freq_odio,
        "frec_no_odio (%)": freq_no_odio.reindex(freq_odio.index).fillna(0),
    }).reset_index().rename(columns={"index": "lema"})

    # Ordenar por frecuencia en odio
    diccionario_odio = diccionario_odio.sort_values("frec_odio (%)", ascending=False).head(200)
    diccionario_odio = diccionario_odio[(diccionario_odio["frec_odio (%)"]    > 0.1) 
                                        & (diccionario_odio["frec_no_odio (%)"] < 0.1)     # raro en no odio
                            ]       
    #print(diccionario_odio)
    return diccionario_odio

def clasificar_por_lemas(texto, lemas_odio):
    doc_nuevo = nlp(texto)
    
    lemas_comentario = [
        token.lemma_.lower()
        for token in doc_nuevo
        if not token.is_stop and not token.is_punct
    ]
    
    coincidencias = set(lemas_comentario) & lemas_odio
    
    return {
        "prediccion":     "odio"    if coincidencias else "no odio",
        "coincidencias":  coincidencias,
        "lemas":          lemas_comentario
    }

diccionario_frecuencias_odio=definir_diccionario_lemas_odio(df_odio, df_no_odio)
print(f"Lemas discriminativos de odio: {len(diccionario_frecuencias_odio)}")
print(diccionario_frecuencias_odio)
print(type(diccionario_frecuencias_odio)) 

lemas_odio_set = set(diccionario_frecuencias_odio["lema"].values)

# Probar con ejemplos manuales
ejemplos = [
    "eres un imbécil y una vergüenza",       # esperado: odio
    "el partido estuvo muy bien jugado",      # esperado: no odio
    "este árbitro es un idiota corrupto",     # esperado: odio
    "mañana juegan en el estadio",            # esperado: no odio
]

for texto in ejemplos:
    resultado = clasificar_por_lemas(texto, lemas_odio_set)
    print(f"\nTexto:         {texto}")
    print(f"Predicción:    {resultado['prediccion']}")
    print(f"Coincidencias: {resultado['coincidencias']}")
    print(f"Lemas:         {resultado['lemas']}")



Lemas discriminativos de odio: 55
                       lema  frec_odio (%)  frec_no_odio (%)
2                      puta         0.6409            0.0730
3                      hijo         0.5793            0.0719
7                   catalán         0.3944            0.0752
8                      asco         0.3944            0.0586
12              catalanisme         0.3205            0.0044
15                   basura         0.2712            0.0741
18                    mujer         0.2588            0.0619
28                    tonto         0.2219            0.0630
31                  gentuza         0.2219            0.0287
37                      vox         0.1972            0.0663
38   l’independentisme         0.1972            0.0033
41                     puto         0.1849            0.0221
42                     culo         0.1849            0.0464
44                   cárcel         0.1726            0.0730
45              delincuente         0.1726         

<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
<p>Si colocaria los lemas, por que hemos extraido entidades, palabras, oraciones, si es femenino o masculin y si la entidad es persona, pero el lema se caracteriza por: El lema es la forma base o canónica de la palabra. Así "corriendo", "corrió", "corren" se cuentan como la misma palabra correr, esto me da un significado y  podria tomar el lema y con una base de palabras que se caracterizen por tener odio en ellas, se puede tomar ya los grupos previos, y obtener una muestra de lemas y con eso crear un algoritmo que genere una puntuación de acuerdo a la coincidencia con los lemas de odio que contiene el corpus.</p>
 